toolkit/tools/pkg/imagecustomizerlib/imagehistory.go

```diff
@@ -0,0 +1,246 @@
+	data, err := json.Marshal(config)
+	if err != nil {
+		return nil, fmt.Errorf("failed to marshal config:\n%w", err)
+	}
+
+	err = json.Unmarshal(data, configCopy)
+	if err != nil {
+		return nil, fmt.Errorf("failed to unmarshal config:\n%w", err)
+	}
+
+	return configCopy, nil
+}
+
+func modifyConfig(configCopy *imagecustomizerapi.Config, baseConfigPath string) error {
+	var err error
+	redactedString := "[redacted]"
+
+	err = populateScriptsList(configCopy.Scripts, baseConfigPath)
+	if err != nil {
+		return fmt.Errorf("failed to populate scripts list:\n%w", err)
+	}
+
+	err = populateAdditionalFiles(configCopy.OS.AdditionalFiles, baseConfigPath)
+	if err != nil {
+		return fmt.Errorf("failed to populate additional files:\n%w", err)
+	}
+
+	err = populateAdditionalDirs(configCopy.OS.AdditionalDirs, baseConfigPath)
+	if err != nil {
+		return fmt.Errorf("failed to populate additional dirs:\n%w", err)
+	}
+
+	err = redactSshPublicKeys(configCopy.OS.Users, redactedString)
+	if err != nil {
+		return fmt.Errorf("failed to redact ssh public keys:\n%w", err)
+	}
+	return nil
+}
+
+func populateAdditionalDirs(configAdditionalDirs imagecustomizerapi.DirConfigList, baseConfigPath string) error {
+	for i := range configAdditionalDirs {
+		hashes := make(map[string]string)
+		sourcePath := configAdditionalDirs[i].Source
+		dirPath := file.GetAbsPathWithBase(baseConfigPath, sourcePath)
+
+		addFileHashToMap := func(path string, d os.DirEntry, err error) error {
+			if err != nil {
+				return err
+			}
+
+			if d.IsDir() {
+				return nil
+			}
+
+			relPath, err := filepath.Rel(dirPath, path)
+			if err != nil {
+				return fmt.Errorf("error computing relative path for %s:\n%w", path, err)
+			}
+
+			hash, err := generateSHA256(path)
+			if err != nil {
+				return err
+			}
+
+			hashes[relPath] = hash
+			return nil
+		}
+
+		err := filepath.WalkDir(dirPath, addFileHashToMap)
+		if err != nil {
+			return fmt.Errorf("error walking directory %s:\n%w", dirPath, err)
+		}
+		configAdditionalDirs[i].SHA256HashMap = hashes
+	}
+	return nil
+}
+
+func populateAdditionalFiles(configAdditionalFiles imagecustomizerapi.AdditionalFileList, baseConfigPath string) error {
+	for i := range configAdditionalFiles {
+		if configAdditionalFiles[i].Source != "" {
+			absSourceFile := file.GetAbsPathWithBase(baseConfigPath, configAdditionalFiles[i].Source)
+			hash, err := generateSHA256(absSourceFile)
+			if err != nil {
+				return err
+			}
+			configAdditionalFiles[i].SHA256Hash = hash
+		}
+	}
+	return nil
+}
+
+func redactSshPublicKeys(configUsers []imagecustomizerapi.User, redactedString string) error {
+	for i := range configUsers {
+		user := configUsers[i]
+		for j := range user.SSHPublicKeys {
+			user.SSHPublicKeys[j] = redactedString
+		}
+
+	}
+	return nil
+}
+
+func populateScriptsList(scripts imagecustomizerapi.Scripts, baseConfigPath string) error {
+	for i := range scripts.PostCustomization {
+		path := scripts.PostCustomization[i].Path
```

Comment:
`PostCustomization` and `FinalizeCustomization` have the same type. So, you can easily de-duplicate the code for them.

Reply:
done